src/GraphODataTemplateWriter/CodeHelpers/Java/TypeHelperJava.cs

```diff
@@ -1165,15 +1185,35 @@ public static string CreatePackageDef(this CustomT4Host host)
 
             // determine current namespace and generate method imports if applicable
             string @namespace;
-            string methodImports = string.Empty;
+            var methodImports = new HashSet<string>();
+            const string importFormat = "import {0}.{1}.{2};";
+            const string graphServiceEntityName = "GraphService";
+            const string interfaceTemplatePrefix = "I";
             switch (host.CurrentType)
             {
                 case OdcmProperty p:
                     @namespace = p.Type.Namespace.Name.AddPrefix();
+                    if (p.Class.GetTypeString() != graphServiceEntityName)
+                        methodImports.Add(string.Format(importFormat, p.Class.Namespace.Name.AddPrefix(), p.Class.GetPackagePrefix(), p.Class.GetTypeString()));
+                    if (!(p.Projection.Type is OdcmPrimitiveType))
+                        methodImports.Add(string.Format(importFormat, p.Projection.Type.Namespace.Name.AddPrefix(), p.Projection.Type.GetPackagePrefix(), p.Projection.Type.GetTypeString()));
+                    p.Projection?.Type?.AsOdcmClass()?.MethodsAndOverloads()?.Distinct()?.SelectMany(o => ImportClassesOfMethodParameters(o))?.ToList()?.ForEach(x => methodImports.Add(x));
                     break;
                 case OdcmMethod m:
-                    var sb = new StringBuilder(Environment.NewLine);
-                    methodImports = ImportClassesOfMethodParameters(m, "import {0}.{1}.{2};", sb).ToString();
+                    m.WithDistinctOverloads().SelectMany(o => ImportClassesOfMethodParameters(o))?.ToList()?.ForEach(x => methodImports.Add(x));
+                    goto default;
+                case OdcmClass c:
+                    if (c.GetTypeString() != graphServiceEntityName)
+                        methodImports.Add(string.Format(importFormat, c.Namespace.Name.AddPrefix(), c.GetPackagePrefix(), c.GetTypeString()));
+
+                    var importTypeToExclude = host.TemplateFile.EndsWith("BaseEntityRequest.java.tt") ? host.TemplateName : string.Empty;
+                    c?.MethodsAndOverloads()?.Distinct()?.SelectMany(o => ImportClassesOfMethodParameters(o, importTypeToExclude: importTypeToExclude))?.ToList()?.ForEach(x => methodImports.Add(x));
+                    c?.NavigationProperties()?.Where(x => x.IsCollection)?.Select(x => x.Projection.Type)?.Distinct()?.ToList()?.ForEach(x =>
+                        ImportRequestBuilderTypes(host, x, methodImports, importFormat, interfaceTemplatePrefix, true)
+                    );
+                    c?.NavigationProperties()?.Where(x => !x.IsCollection)?.Select(x => x.Projection.Type)?.Distinct()?.ToList()?.ForEach(x =>
```

Comment:
nit: style: I find reading layered linq queries more readable when each function is in a different line (feels like SQL queries that way).

Reply:
fixed